### ISSUE TITLE
Resource embedding depends on view naming

### ISSUE THREAD

**Issue by nickmyatt** [begriffs/postgrest#823]
OS

```
$ lsb_release -ar
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04 LTS
Release:	16.04
Codename:	xenial
```

PostgREST version

```
$ ./postgrest |& grep PostgREST
PostgREST 0.4.0.0 / create a REST API to an existing Postgres database
```

With the following database

```
CREATE SCHEMA app;
CREATE TABLE app.authors (id serial PRIMARY KEY, name text);
CREATE TABLE app.books (title text, author_id integer REFERENCES app.authors (id));

INSERT INTO app.authors (id, name) VALUES (1, 'Steinbeck');
INSERT INTO app.books (title, author_id) VALUES ('The Grapes of Wrath', 1)

CREATE SCHEMA app_api;
CREATE VIEW app_api.authors AS (SELECT * FROM app.authors);
CREATE VIEW app_api.books AS (SELECT * FROM app.books);
CREATE VIEW app_api.literary_works AS (SELECT * FROM app.books);
```

After starting postgrest with `db-schema = "app_api"`, I get what I expect from the `/books` endpoint

```
$ curl 'localhost:3000/books?select=*,author\{*\}'
[{"title":"The Grapes of Wrath","author_id":1,"author":{"id":1,"name":"Steinbeck"}}]
```

But not from `/literary_works`

```
$ curl "localhost:3000/literary_works?select=*,author\{*\}"
{"message":"could not find foreign keys between these entities, no relation between literary_works and author"}
```

I would expect resource embedding to work the in the same way for trivial views such as this one (I realise it does not work for views with subselects).

**Comment by begriffs** [comment]
@ruslantalpa you're the resource embedding guy, how hard do you think this is to fix?

**Comment by ruslantalpa** [comment]
i have not checked this is actually happening since i remember we have this
https://github.com/begriffs/postgrest/blob/master/test/Feature/QuerySpec.hs#L273-L279

so it's not the name of the view

**Comment by nickmyatt** [comment]
I just discovered that this only happens when the view and table belong to different schemas.

So, if I were to instead define `CREATE VIEW app.literary_works AS ...` (rather than in `app_api` schema) the resource embedding behaves as I expect.

So, as a workaround I can move *all* of my views to the main `app` schema and point PostgREST at that. However, when it comes time to release a new API version, I am going to need a separate schema for the  `v2` views, aren't I?

**Comment by begriffs** [comment]
Yeah you will. Each version would live in a different schema, and you would spin up a postgrest instance for each, configured to use the desired schema and running on separate ports. Then use nginx to route requests to the correct postgrest instance (based on their accept header, or url prefix or however you like to expose versions over http).

**Comment by nickmyatt** [comment]
If I run [the query PostgREST uses to identify view columns](https://gist.github.com/ruslantalpa/2eab8c930a65e8043d8f) I get

```
 table_schema | table_name | table_column_name | view_schema |   view_name    | view_column_name 
--------------+------------+-------------------+-------------+----------------+------------------
 app          | authors    | id                | app_api     | authors        | id
 app          | authors    | name              | app_api     | authors        | name
 app          | books      | author_id         | app_api     | books          | author_id
 app          | books      | title             | app_api     | books          | title
 app          | books      | author_id         | app_api     | literary_works | author_id
 app          | books      | title             | app_api     | literary_works | title
```

i.e. the same results for `app_api.books` and `app_api.literary_works`, so it seems the problem is elsewhere.

**Comment by begriffs** [comment]
I wonder if this has anything to do with the schema search_path. PR #831 recently dealt with a search_path related thing...

**Comment by umq** [comment]
in the example, dropping `app_api.books` makes the latter query `/literary_works?select=*,author{*}` okay.
There may be something like the-first-one-wins rule.

**Comment by umq** [comment]
replacing ` = 1` here to ` > 0` possibly work?
https://github.com/calebmer/postgrest/blob/master/src/PostgREST/DbStructure.hs#L332

**Comment by begriffs** [comment]
@umq what was the meaning of the blockquote involving `=1` vs `>0`?

**Comment by begriffs** [comment]
Yep, tried this example with postgrest built from latest master (7cc3833) and verified it's still broken as described, and that deleting the app_api.books view makes the literary_works embedding start working again. Thanks for the simple and reproducible case @nickmyatt.

Also the `allSynonyms` function has a quite a different form nowadays than it did in the linked code.

**Comment by ruslantalpa** [comment]
Could it be that the views in question are related in two different ways and you are hitting that problme and the name is not a problem?
the reason i think that is because it starts to work when books view is dropped.

i added a pr recently https://github.com/begriffs/postgrest/pull/918 that deals with specifying the exact fk's to use when going the embedding.

can you try the same query with all the views but this time specify the FK to use like the PR explains?

**Comment by ruslantalpa** [comment]
please reopen if the fix added in #918 does not work (it's included in 4.3)
